File: code/unpack.cpp
/*
Copyright (C) 2003 Parallel Realities

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/

#include "unpack.h"

/*
Searches the pak file for the required data. When
it is found, the data is read into a character buffer.
*/
void unpack(char *file, signed char fileType)
{
	unsigned char *packBuffer;
	char packFilename[60];
	int packFSize;

	FILE *pak;
	FILE *fp = NULL; // music has to be read-written-read!
	char musicFilename[PATH_MAX];

	strcpy(packFilename, "");

	pak = fopen(PACKLOCATION, "rb");
	if (pak == NULL)
	{
		printf("Couldn't access the Project: Starfighter data file!\n");
		exit(1);
	}
	fseek(pak, 4, SEEK_SET);

	while (true)
	{
		if (!fread(packFilename, 1, 56, pak))
		{
			fclose(pak);
			if (fileType != PAK_FONT)
			{
				showErrorAndExit(0, file);
			}
			exit(1);
			break;
		}

		fread(&packFSize, 4, 1, pak);
		packFSize = SDL_SwapLE32(packFSize);

		if (strcmp(packFilename, file) == 0)
		{
			if ((fileType == PAK_MOD) || (fileType == PAK_S3M))
			{
				if (fileType == PAK_MOD)
				{
					sprintf(musicFilename, "%smusic.mod", engine.userHomeDirectory);
					fp = fopen(musicFilename, "wb");
				}
				else
				{
					sprintf(musicFilename, "%smusic.s3m", engine.userHomeDirectory);
					fp = fopen(musicFilename, "wb");
				}

				if (fp == NULL)
					showErrorAndExit(1, "");
			}
			packBuffer = (unsigned char*) malloc(packFSize);
			fread(packBuffer, 1, packFSize, pak);
			if ((fileType == PAK_MOD) || (fileType == PAK_S3M))
			{
				fwrite(packBuffer, 1, packFSize, fp);
				fclose(fp);
			}
			break;
		}
		else
		{
			fseek(pak, packFSize, SEEK_CUR);
		}
	}

	if ((fileType != PAK_MOD) && (fileType != PAK_S3M))
		engine.sdlrw = SDL_RWFromMem(packBuffer, packFSize);

	//printf("Extracted: %s\n", file);

	fclose(pak);
}

/*
Search the data package for the required file.
When it is found, return the location.
*/
int locateDataInPak(char *file, signed char required)
{
	//printf("Looking for %s...", file);

	char packFilename[60];
	int packFSize;
	int location = 0;

	FILE *pak;

	strcpy(packFilename, "");

	pak = fopen(PACKLOCATION, "rb");
	if (pak == NULL)
	{
		printf("Couldn't access the Project: Starfighter data file!\n");
		exit(1);
	}
	fseek(pak, 4, SEEK_SET);

	while (true)
	{
		if (!fread(packFilename, 1, 56, pak))
		{
			fclose(pak);
			if (required)
			{
				showErrorAndExit(0, file);
				exit(1);
			}
			break;
		}
		fread(&packFSize, 4, 1, pak);
		packFSize = SDL_SwapLE32(packFSize);

		if (strcmp(packFilename, file) == 0)
		{
			location = ftell(pak);
			fclose(pak);

			//printf("found it!\n");

			return location;
		}
		else
		{
			fseek(pak, packFSize, SEEK_CUR);
		}
	}

	//printf("not found (skipping)\n");

	return -1; // we only get this if it isn't required
}

File: unpack.py
#!/usr/bin/env python
# unpack.py - unpack files from a Parallel Realities Starfighter .pak file
# Copyright (C) 2006 Josh Triplett <josh@freedesktop.org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License only.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

import os, struct, sys

class PakException(Exception): pass

class BadMagic(PakException):
    def __init__(self, pak):
        PakException.__init__(self,
            '%s: not a pack file (does not start with "PACK")' % pak.name)

class ShortRead(PakException):
    def __init__(self, pak):
        PakException.__init__(self, '%s: file corrupt (short read)' % pak.name)

class PathTraversal(PakException):
    def __init__(self, pak, filename):
        PakException.__init__(self,
            '%s: attempted directory traversal with filename "%s"'
            % (pak.name, filename))

def full_read(pak, size):
    data = pak.read(size)
    if len(data) != size: raise ShortRead(pak)
    return data

def unpack(pak):
    pwd = os.getcwd()
    if full_read(pak, 4) != "PACK": raise BadMagic(pak)
    while True:
        filename = pak.read(56)
        if filename == "": break # End of file
        elif len(filename) < 56: raise ShortRead(pak)
        filename = filename.split("\x00")[0]
        (size,) = struct.unpack("<I", full_read(pak, 4))
        data = full_read(pak, size)
        dirname = os.path.abspath(os.path.dirname(filename))
        if(not(dirname.startswith(pwd))):
            raise PathTraversal(pak, filename)
        try: os.makedirs(dirname)
        except OSError, e:
            if e.errno != 17: raise # Ignore "file exists"
        out = file(filename, "w+")
        out.write(data)
        out.close()

def main(args):
    try:
        pak = args[1]
        return unpack(file(pak, "rb"))
    except IndexError:
        print "Usage: unpack filename.pak"
    except (IOError, OSError, PakException), e:
        print e
    return 1

__all__ = ["unpack", "PakException", "BadMagic", "ShortRead", "PathTraversal"]

if __name__ == "__main__": sys.exit(main(sys.argv))
